src/coreclr/System.Private.CoreLib/src/System/Array.CoreCLR.cs

```diff
@@ -95,8 +97,289 @@ private static unsafe void CopyImpl(Array sourceArray, int sourceIndex, Array de
-        private static extern void CopySlow(Array sourceArray, int sourceIndex, Array destinationArray, int destinationIndex, int length);
+        private static unsafe void CopySlow(Array sourceArray, int sourceIndex, Array destinationArray, int destinationIndex, int length)
+        {
+            Debug.Assert(sourceArray.Rank == destinationArray.Rank);
+
+            TypeHandle srcTH = RuntimeHelpers.GetMethodTable(sourceArray)->GetArrayElementTypeHandle();
+            TypeHandle destTH = RuntimeHelpers.GetMethodTable(destinationArray)->GetArrayElementTypeHandle();
+            AssignArrayEnum r = CanAssignArrayType(srcTH.m_asTAddr, destTH.m_asTAddr);
+
+            if (r == AssignArrayEnum.AssignWrongType)
+                ThrowHelper.ThrowArrayTypeMismatchException_CantAssignType();
+
+            if (length > 0)
+            {
+                switch (r)
+                {
+                    case AssignArrayEnum.AssignUnboxValueClass:
+                        UnBoxEachElement(sourceArray, sourceIndex, destinationArray, destinationIndex, length);
+                        break;
+
+                    case AssignArrayEnum.AssignBoxValueClassOrPrimitive:
+                        BoxEachElement(sourceArray, sourceIndex, destinationArray, destinationIndex, length);
+                        break;
+
+                    case AssignArrayEnum.AssignMustCast:
+                        CastCheckEachElement(sourceArray, sourceIndex, destinationArray, destinationIndex, length);
+                        break;
+
+                    case AssignArrayEnum.AssignPrimitiveWiden:
+                        PrimitiveWiden(sourceArray, sourceIndex, destinationArray, destinationIndex, length);
+                        break;
+
+                    default:
+                        Debug.Fail("Fell through switch in Array.Copy!");
+                        break;
+                }
+            }
+        }
+
+        private enum AssignArrayEnum
+        {
+            AssignWrongType,
+            AssignMustCast,
+            AssignBoxValueClassOrPrimitive,
+            AssignUnboxValueClass,
+            AssignPrimitiveWiden,
+        }
+
+        [LibraryImport(RuntimeHelpers.QCall, EntryPoint = "Array_CanAssignArrayType")]
+        private static unsafe partial AssignArrayEnum CanAssignArrayType(void* srcTH, void* dstTH);
+
+        // Unboxes from an Object[] into a value class or primitive array.
+        private static unsafe void UnBoxEachElement(Array sourceArray, int sourceIndex, Array destinationArray, int destinationIndex, int length)
+        {
+            MethodTable* pDestArrayMT = RuntimeHelpers.GetMethodTable(destinationArray);
+            TypeHandle destTH = pDestArrayMT->GetArrayElementTypeHandle();
+            // _ASSERTE(destTH.GetSignatureCorElementType() == ELEMENT_TYPE_CLASS || destTH.GetSignatureCorElementType() == ELEMENT_TYPE_VALUETYPE || CorTypeInfo::IsPrimitiveType(pDest->GetArrayElementType()));
```

Comment:
What does ELEMENT_TYPE_CLASS mean for this case?

Reply:
I do not think it makes sense to allow ELEMENT_TYPE_CLASS here.